runtime/src/accounts_db.rs

```diff
@@ -1724,7 +1724,9 @@ impl AccountsDb {
         // Don't reset from clean, since the pubkeys in those stores may need to be unref'ed
         // and those stores may be used for background hashing.
         let reset_accounts = false;
-        self.handle_reclaims(&reclaims, None, false, None, reset_accounts);
+        let mut reclaim_result = ReclaimResult::default();
+        let reclaim_result = Some(&mut reclaim_result);
+        self.handle_reclaims(&reclaims, None, false, reclaim_result, reset_accounts);
```

Comment:
ah shoot, this is because the final `finalize_dead_slot_removal` only unrefs the pubkey if `purged_stored_account_slot.is_some()`. This seems like the likely bug.

Looks like I introduced it here: https://github.com/solana-labs/solana/pull/15534/files#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR3958, ugh sorry!

Reply:
A good thing to enforce in `handle_reclaims()` might be that anytime that `no_dead_slot == false,`, then `reclaim_result.is_some()` should also be true, because if a slot can be dead, then unrefing must be able to happen.

Reply:
Yea, this seems right. Good catch @jeffwashington

Reply:
#15534 is the introduction of this problem.
Unless I'm mistaken, I see this commit in 1.5 and 1.6.
@mvines @sakridge  what do you recommend?